examples/fluids/problems/shocktube.c

```diff
@@ -0,0 +1,246 @@
+  second = fabs(second);
+
+  ierr = PetscOptionsEnd(); CHKERRQ(ierr);
+
+  // ------------------------------------------------------
+  //           Set up the PETSc context
+  // ------------------------------------------------------
+  user->units->meter  = meter;
+  user->units->second = second;
+
+  // ------------------------------------------------------
+  //           Set up the libCEED context
+  // ------------------------------------------------------
+  // -- Scale variables to desired units
+  lx = fabs(lx) * meter;
+  ly = fabs(ly) * meter;
+  lz = fabs(lz) * meter;
+  CeedScalar midpt = 0.5*lx;
+
+  // -- Setup Context
+  setup_context->lx        = lx;
+  setup_context->ly        = ly;
+  setup_context->lz        = lz;
+  setup_context->midpt     = midpt;
+  setup_context->time      = 0.0;
+  setup_context->P_high    = P_high;
+  setup_context->rho_high  = rho_high;
+  setup_context->P_low     = P_low;
+  setup_context->rho_low   = rho_low;
+
+  // -- QFunction Context
+  user->phys->implicit                      = implicit;
+  user->phys->has_curr_time                 = has_curr_time;
+  user->phys->shocktube_ctx->implicit       = implicit;
+  user->phys->shocktube_ctx->stabilization  = stab;
+  user->phys->shocktube_ctx->yzb            = yzb;
+  user->phys->shocktube_ctx->cv             = cv;
+  user->phys->shocktube_ctx->Cyzb           = Cyzb;
+  user->phys->shocktube_ctx->Byzb           = Byzb;
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode SetupContext_SHOCKTUBE(Ceed ceed, CeedData ceed_data,
+                                      AppCtx app_ctx, SetupContext setup_ctx, Physics phys) {
+  PetscFunctionBeginUser;
+
+  CeedQFunctionContextCreate(ceed, &ceed_data->setup_context);
+  CeedQFunctionContextSetData(ceed_data->setup_context, CEED_MEM_HOST,
+                              CEED_USE_POINTER, sizeof(*setup_ctx), setup_ctx);
+  CeedQFunctionSetContext(ceed_data->qf_ics, ceed_data->setup_context);
+  CeedQFunctionContextCreate(ceed, &ceed_data->shocktube_context);
+  CeedQFunctionContextSetData(ceed_data->shocktube_context, CEED_MEM_HOST,
+                              CEED_USE_POINTER,
+                              sizeof(*phys->shocktube_ctx), phys->shocktube_ctx);
+  if (ceed_data->qf_rhs_vol)
+    CeedQFunctionSetContext(ceed_data->qf_rhs_vol, ceed_data->shocktube_context);
+  if (ceed_data->qf_ifunction_vol)
+    CeedQFunctionSetContext(ceed_data->qf_ifunction_vol,
+                            ceed_data->shocktube_context);
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode BC_SHOCKTUBE(DM dm, SimpleBC bc, Physics phys,
+                            void *setup_ctx) {
+  PetscErrorCode ierr;
+  PetscFunctionBeginUser;
+
+  // Default boundary conditions
+  //    slip BC on y and z faces
+  bc->num_slip[0] = 0;
+  bc->num_slip[1] = bc->num_slip[2] = 4;
+  bc->slips[1][0] = bc->slips[2][0] = 1;
+  bc->slips[1][1] = bc->slips[2][1] = 2;
+  bc->slips[1][2] = bc->slips[2][2] = 3;
+  bc->slips[1][3] = bc->slips[2][3] = 4;
```

Comment:
@tt-aiken , I don't understand how the BCs are specified. If you want to set slip BCs on y and z directions (based on your comment), the following should do the job.

```
  bc->num_slip[1] = bc->num_slip[2] = 2;
  bc->slips[1][0] = 3;
  bc->slips[1][1] = 4;
  bc->slips[2][0] = 1;
  bc->slips[2][1] = 2;
```

Is there a reason you have considered 4 faces for each direction?

Reply:
I see that setting slip BC for y-velocity on the y-face is redundant, and same with z-velocities on z-faces - two faces should do the trick. However, from my understanding of the face numbering convention, this would look like:

```
bc->num_slip[1] = bc->num_slip[2] = 2;
bc->slips[1][0] = 1;
bc->slips[1][1] = 2;
bc->slips[2][0] = 3;
bc->slips[2][1] = 4;
```

These PETSc BC assignment routines are a bit opaque to me though. Am I missing something here?

Reply:
Right, it took me a while to understand the numbering convention. Hmm, faces `1` and `2` are in the `z` direction, `3` and `4` in the `y` direction, and `5` and `6` in the `x` direction. So, the one that I wrote above should be correct. Just like what we have in [densitycurrent](https://github.com/CEED/libCEED/blob/239c65026b8097e4437d6328197c2f08879797b1/examples/fluids/problems/densitycurrent.c#L283).

Reply:
The numbering is from here: https://gitlab.com/petsc/petsc/-/blob/main/src/dm/impls/plex/plexinterpolate.c#L177 . The z faces come first so that the closure of an interpolated mesh produces the same vertex order as the uninterpolated mesh.

Reply:
Thank you both, I'll change this now.